bkbit/scripts/gfftranslator.py

```diff
@@ -1,203 +1,682 @@
-        for g in annotations:
-            output_arr.append(g.dict(exclude_none=True))
-        f.write(json.dumps(output_arr, indent=2))
-
-def gff_to_gene_annotation(input_fname, data_dir, output_dir):
-    """Converts GFF file(s) to GeneAnnotation objects and serializes them to a JSON file.
-
-    Parameters:
-    gff_files_path (str): path .csv that contains GFF file(s)
-    output_dir (str): path to output directory
-
-    """
-    orgTaxonObjects = {}
-    genomeAnnotObjects = {}
-    if not os.path.exists(output_dir):
-        os.makedirs(output_dir)
-    with open(os.path.join(data_dir,input_fname), newline='') as csvfile:
-        reader = csv.DictReader(csvfile)
-        for row in reader:
-            print(f"AUTHORITY: {row['authority']}, LABEL: {row['label']}, TAXON_LOCAL_UNIQUE_ID: {row['taxon_local_unique_identifier']}, VERSION: {row['version']}, GENE_ID_PREFIX: {row['gene_identifier_prefix']}, URL: {row['url']}")
-            gene_name = ''.join(row['url'].split('/')[-1].split('.')[0:-2])
-            print("GENE NAME: ", gene_name)
-            if os.path.isfile(data_dir + '/' + gene_name + '.csv'):
-                print(f"Data from url is already downloaded and saved here: {data_dir + '/' + gene_name + '.csv'}")
-                df = pd.read_csv(data_dir + '/' + gene_name + '.csv') 
+
+logging.basicConfig(
+    filename="gff3_translator_" + datetime.now().strftime("%Y-%m-%d_%H:%M:%S") + ".log",
+    format="%(levelname)s: %(message)s (%(asctime)s)",
+    datefmt="%m/%d/%Y %I:%M:%S %p",
+    level=logging.INFO,
+)
+logger = logging.getLogger(__name__)
+
+
+## CONSTANTS ##
+
+TAXON_SCIENTIFIC_NAME = {
+    "9606": "Homo sapiens",
+    "10090": "Mus musculus",
+    "9544": "Macaca mulatta",
+    "9483": "Callithrix jacchus",
+    "60711": "Chlorocebus sabaeus",
+}
+TAXON_COMMON_NAME = {
+    "9606": "human",
+    "10090": "mouse",
+    "9544": "rhesus macaque",
+    "9483": "common marmoset",
+    "60711": "green monkey",
+}
+PREFIX_MAP = {
+    "NCBITaxon": "http://purl.obolibrary.org/obo/NCBITaxon_",
+    "NCBIGene": "http://identifiers.org/ncbigene/",
+    "ENSEMBL": "http://identifiers.org/ensembl/",
+    "NCBIAssembly": "https://www.ncbi.nlm.nih.gov/assembly/",
+}
+NCBI_GENE_ID_PREFIX = "NCBIGene"
+ENSEMBL_GENE_ID_PREFIX = "ENSEMBL"
+TAXON_PREFIX = "NCBITaxon"
+ASSEMBLY_PREFIX = "NCBIAssembly"
+BICAN_ANNOTATION_PREFIX = "bican:annotation-"
+GENOME_ANNOTATION_DESCRIPTION_FORMAT = (
+    "{authority} {taxon_scientific_name} Annotation Release {genome_version}"
+)
+DEFAULT_FEATURE_FILTER = ("gene", "pseudogene", "ncRNA_gene")
+DEFAULT_HASH = ("SHA256", "MD5")
+
+
+class Gff3:
+    def __init__(
+        self,
+        content_url,
+        taxon_id,
+        assembly_id,
+        assembly_version,
+        assembly_label,
+        genome_label: str,
+        genome_version: str,
+        genome_authority: str,
+        hash_functions: tuple[str] = DEFAULT_HASH,
+        assembly_strain=None,
+        gff_file=None,
```

Comment:
The content_url is required because it is an attribute in GenomeAnnotation. So if a user only provides a content_url I will download the gff3 file. And if a user provides a gff_file then I will use the provided file to create the gene annotations and only use the provided content_url to set the value of GenomeAnnotation.content_url.